### PR TITLE
LibWeb: Make CSSPixels constexpr and do some more rounding and saturation

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
### LibWeb: Truncate on `to_int` in PixelUnits

This matches what floating point numbers do on casting.
Previous behavior was a `floor` operation

### LibWeb: Use CSSPixels for CSSPixels::epsilon

It was only ever used to compare to CSSPixels, so let's avoid some casts

### CODE REVIEW

File: Tests/LibWeb/Layout/expected/table/percentage-width-columns.txt [hunk 1]
```diff
@@ -1,33 +1,33 @@
 Viewport <#document> at (0,0) content-size 800x600 children: not-inline
   BlockContainer <html> at (0,0) content-size 800x600 [BFC] children: not-inline
-    BlockContainer <body> at (8,8) content-size 784x22.46875 children: not-inline
-      TableWrapper <(anonymous)> at (8,8) content-size 88.328125x22.46875 [BFC] children: not-inline
```
Comment (kalenikaliaksandr) [inline Tests/LibWeb/Layout/expected/table/percentage-width-columns.txt:4]: `88.328125` -> `93.328125` is more difference then I would expect to see considering the changes. regression?

File: Tests/LibWeb/Layout/expected/table/percentage-width-columns.txt [hunk 2]
```diff
@@ -1,33 +1,33 @@
...
-                  frag 0 from TextNode start: 0, length: 1, rect: [47.484375,10.5 9.34375x17.46875]
+                  frag 0 from TextNode start: 0, length: 1, rect: [50.265625,11 9.34375x17.46875]
                     "B"
                 TextNode <#text>
               BlockContainer <(anonymous)> (not painted) children: inline
                 TextNode <#text>
-              BlockContainer <td> at (79.5625,10.5) content-size 14.265625x17.46875 table-cell [BFC] children: inline
+              BlockContainer <td> at (84.0625,11) content-size 14.265625x17.46875 table-cell [BFC] children: inline
                 line 0 width: 10.3125, height: 17.46875, bottom: 17.46875, baseline: 13.53125
-                  frag 0 from TextNode start: 0, length: 1, rect: [81.53125,10.5 10.3125x17.46875]
```
Comment (kalenikaliaksandr) [inline Tests/LibWeb/Layout/expected/table/percentage-width-columns.txt:30]: same here.

File: Tests/LibWeb/TestCSSPixels.cpp [hunk 3]
```diff
@@ -87,4 +87,40 @@ TEST_CASE(saturated_subtraction)
     EXPECT_EQ(value - -1, CSSPixels(INFINITY));
 }
 
+TEST_CASE(rounding)
```
Comment (kalenikaliaksandr) [inline Tests/LibWeb/TestCSSPixels.cpp:90]: tests for rounding is nice! maybe some tests for changes in `*` and `/` would be even better :)

**Comment by kalenikaliaksandr** [comment]
There are some suspicious changes in layout test dumps but otherwise looks good to merge!

**Comment by Hendiadyoin1** [comment]
So I have checked by calling into LibM instead and the tests dont seem to change

<details><summary>diff</summary>
<p>

```diff
diff --git a/Userland/Libraries/LibWeb/PixelUnits.h b/Userland/Libraries/LibWeb/PixelUnits.h
index 6b4ec64651..e60b8ff304 100644
--- a/Userland/Libraries/LibWeb/PixelUnits.h
+++ b/Userland/Libraries/LibWeb/PixelUnits.h
@@ -254,24 +254,19 @@ inline Web::CSSPixels abs(Web::CSSPixels const& value)
     return value.abs();
 }
 
-constexpr Web::CSSPixels floor(Web::CSSPixels const& value)
+inline Web::CSSPixels floor(Web::CSSPixels const& value)
 {
-    return Web::CSSPixels::from_raw(value.raw_value() & ~Web::CSSPixels::radix_mask);
+    return floor(value.to_double());
 }
 
-constexpr Web::CSSPixels ceil(Web::CSSPixels const& value)
+inline Web::CSSPixels ceil(Web::CSSPixels const& value)
 {
-    auto floor_value = value.raw_value() & ~Web::CSSPixels::radix_mask;
-    auto ceil_value = floor_value + (value.raw_value() & Web::CSSPixels::radix_mask ? Web::CSSPixels::fixed_point_denominator : 0);
-    return Web::CSSPixels::from_raw(ceil_value);
+    return ceil(value.to_double());
 }
 

-constexpr Web::CSSPixels round(Web::CSSPixels const& value)
+inline Web::CSSPixels round(Web::CSSPixels const& value)
 {
-    // FIXME: Maybe do this with bit-fiddling instead
-    if (value > 0)
-        return floor(value + Web::CSSPixels::from_raw(Web::CSSPixels::fixed_point_denominator >> 1 /* 0.5 */));
-    return ceil(value - Web::CSSPixels::from_raw(Web::CSSPixels::fixed_point_denominator >> 1 /* 0.5 */));
+    return round(value.to_double());
 }
 
 constexpr Web::DevicePixels abs(Web::DevicePixels const& value)
```

</p>
</details> 

so no clue why the changes are rather big

File: Tests/LibWeb/TestCSSPixels.cpp [hunk 4]
```diff
@@ -31,6 +31,11 @@ TEST_CASE(division1)
     CSSPixels b(5);
     CSSPixels c = a / b;
     EXPECT_EQ(c, CSSPixels(2));
+
```
Comment (kalenikaliaksandr) [inline Tests/LibWeb/TestCSSPixels.cpp:34]: typo in commit message: "LibWeb: Saturare result in PixelUnits::operator/". Saturare -> Saturate.

File: Userland/Libraries/LibWeb/PixelUnits.cpp [hunk 5]
```diff
@@ -14,19 +14,33 @@ static constexpr i32 fixed_point_denominator = 1 << fractional_bits;
...
+
 CSSPixels::CSSPixels(int value)
 {
-    m_value = value * fixed_point_denominator;
+    if (value > max_integer_value) [[unlikely]]
+        m_value = NumericLimits<int>::max();
+    else if (value < min_integer_value) [[unlikely]]
+        m_value = NumericLimits<int>::min();
+    else
+        m_value = static_cast<int>(value) << fractional_bits;
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/PixelUnits.cpp]: `static_cast` is redundant here.

File: Userland/Libraries/LibWeb/PixelUnits.cpp [hunk 6]
```diff
@@ -14,19 +14,33 @@ static constexpr i32 fixed_point_denominator = 1 << fractional_bits;
...
 }
 
 CSSPixels::CSSPixels(unsigned long value)
 {
-    m_value = value * fixed_point_denominator;
+    if (value > max_integer_value) [[unlikely]]
+        m_value = NumericLimits<int>::max();
+    else
+        m_value = static_cast<int>(value) << fractional_bits;
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/PixelUnits.cpp]: seems like `static_cast` is not needed here.

File: Userland/Libraries/LibWeb/PixelUnits.cpp [hunk 7]
```diff
@@ -67,7 +67,11 @@ double CSSPixels::to_double() const
 
 int CSSPixels::to_int() const
 {
-    return m_value / fixed_point_denominator;
+    // Truncate on cast
+    return (m_value >> fractional_bits)
+        + ((m_value & radix_mask)
+                ? m_value > 0 ? 0 : 1
+                : 0);
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/PixelUnits.cpp:28]: hmm, could you provide some explanation on why truncation is needed here?

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/PixelUnits.cpp:28]: or better write test that demonstrates why it is required :)

File: Userland/Libraries/LibWeb/PixelUnits.cpp (hunk 6 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibWeb/PixelUnits.cpp]: Sound weird, but I can remove that

File: Userland/Libraries/LibWeb/PixelUnits.cpp (hunk 7 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibWeb/PixelUnits.cpp:28]: I think the commit title says it, but I can add a comment
but essentially that matches what floats do when you cast them
Will also add a test

File: Userland/Libraries/LibWeb/PixelUnits.cpp (hunk 6 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibWeb/PixelUnits.cpp]: grafik

Comment (Hendiadyoin1) [reply Userland/Libraries/LibWeb/PixelUnits.cpp]: Clang-tidy says no

File: Userland/Libraries/LibWeb/PixelUnits.cpp (hunk 7 above)
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/PixelUnits.cpp:28]: it doesn't seem like there any reason why would want to match this truncation behaviour that happens for float.

File: Userland/Libraries/LibWeb/PixelUnits.cpp (hunk 6 above)
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/PixelUnits.cpp]: oh, my mistake. thank you for checking!

**Review by kalenikaliaksandr — APPROVED** [review]

**Pull request merged**